server/src/main/java/org/elasticsearch/health/node/selection/HealthNodeTaskExecutor.java

```diff
@@ -136,7 +138,8 @@ protected HealthNode createTask(
     public PersistentTasksCustomMetadata.Assignment getAssignment(
         HealthNodeTaskParams params,
         Collection<DiscoveryNode> candidateNodes,
-        ClusterState clusterState
+        ClusterState clusterState,
+        @Nullable ProjectId projectId
```

Comment:
We'll probably want to split this method up into a cluster-scoped version and a project-scoped version. The health node persistent task is cluster-scoped, so it doesn't really make sense to have a project ID here or in other cluster-scoped persistent tasks (even though it's nullable). Let me know if "cluster-scoped vs. project-scoped persistent tasks" sound unfamiliar to you, then I (or Yang) can explain what they are. But I'll let @ywangd decide whether he agrees or whether he's fine with the nullable project ID like this.

If we decide to split it up, we can one method without a project ID (for cluster-scoped tasks) and one with a `ProjectState` (instead of a `ClusterState` and `ProjectId`). We created `ProjectState` to avoid passing cluster states together with project IDs.

Reply:
Yeah we can go with two separate methods. The `PersistentTasksExecutor#scope` method can be used to tell the scope of the executor and subsequently call the relevant `getXxxAssignment` method. 

Theoretically we can have a single overriden generic method for project and cluster scoped task executors, if `ProjectState` and `ClusterState` shares some interface, e.g. `Supplier<ClusterState>`. It should help reducing verbosity of the types. We will still need to check the task executor types and pass either `ClusterState` or `ProjectState` to the method accordingly. This might be something worth doing in future since it feels like a better type system. But it is definitely **outside of** this PR.